src/StackExchange.Redis/ServerEndPoint.cs

```diff
@@ -77,6 +76,32 @@ public ServerEndPoint(ConnectionMultiplexer multiplexer, EndPoint endpoint)
 
         public bool IsConnecting => interactive?.IsConnecting == true;
 
+        private readonly List<TaskCompletionSource<bool>> _pendingConnectionMonitors = new List<TaskCompletionSource<bool>>();
+
+        /// <summary>
+        /// Awaitable state seeing if this endpoint is connected.
+        /// </summary>
+        public Task<bool> OnConnectedAsync(LogProxy log = null, bool sendTracerIfConnected = false)
+        {
+            if (!IsConnected)
+            {
+                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Just as commentary, kind of scary to me how the RunContinuationsAsynchronously makes the lock statements far away  do the right thing. But necessary yeah.